### PR TITLE
Introduce new command overload system

### PR DESCRIPTION
**Pull request by alvin0319**
## Introduction

This pull request will rewrite some command execution processes and will provide developers with a more accurate and diverse way to execute the commands.

### Relevant issues

There are no relevant issues.

### API changes

- The following classes have been added:
  - `Overload`
  - `Parameter`
  - `EnumParameter`
  - `PlayerParameter`
  - `StringParameter`
  - `TextParameter`
  - `Vector3Parameter`
  - `NumericParameter`
  - `FloatParameter`
  - `IntegerParameter`
- The following methods have been added:
  - `Command->addOverload()`
  - `Command->getOverloads()`
  - `Command->parseArguments()`
- The following methods have been changed:
  - `Command->execute()` is no longer abstract

### Behavioural changes

There are no behavior changes.

## Backwards compatibility

There are no backwards compatibility issues.

Developers can still override `Command->execute()` and there is no problem with using it.

## Tests

- PHPStan analyze test

```
Note: Using configuration file C:\Users\user\Desktop\MCBE\PocketMine-MP-command-overload\phpstan.neon.dist.
 1282/1282 [============================] 100%

 ------ ---------------------------------------------------------------
  Line   src\command\Overload.php                                      
 ------ ---------------------------------------------------------------
  41     Anonymous function should have native return typehint "void".
  109    Anonymous function should have native return typehint "void".
 ------ ---------------------------------------------------------------

                                                                                                                        
 [ERROR] Found 2 errors
```

- Example plugin

```php
<?php

/**
 * @name TestCommandPlugin
 * @author alvin0319
 * @main alvin0319\TestCommandPlugin\TestCommandPlugin
 * @version 1.0.0
 * @api 4.0.0
 */

declare(strict_types=1);

namespace alvin0319\TestCommandPlugin;

use pocketmine\command\Command;
use pocketmine\command\CommandSender;
use pocketmine\command\Overload;
use pocketmine\command\parameter\defaults\PlayerParameter;
use pocketmine\command\parameter\defaults\Vector3Parameter;
use pocketmine\plugin\PluginBase;

class TestCommandPlugin extends PluginBase{

	protected function onEnable() : void{
		$this->getServer()->getCommandMap()->register("test", new class extends Command{

			public function __construct(){
				parent::__construct("test", "Test command");
				$this->addOverload((new Overload())
					->addParameter((new Vector3Parameter("vector")))
					->setCommandHandler(function(CommandSender $sender, array $args){
						$vector = $args["vector"];

						$sender->sendMessage("Vector: $vector");
					})
				);
				$this->addOverload((new Overload(function(CommandSender $sender, array $args){
					$target = $args["target"]->getName();
					$sender->sendMessage("Target: {$target}");
				}))
					->addParameter(new PlayerParameter("target"))
				);
			}
		});
	}
}
```

### CODE REVIEW

**Comment by dktapps** [comment]
It would be preferable to have callbacks bound to each overload instead of a central `onRun()`. That's how I'd personally planned to implement this.

File: src/command/Overload.php [hunk 1]
```diff
@@ -0,0 +1,107 @@
...
+
+class Overload{
+	/** @var Command */
+	protected $command;
+
+	/** @var Parameter[] */
+	protected $parameters = [];
+
+	public function __construct(Command $command){
+		$this->command = $command;
```
Comment (dktapps) [inline src/command/Overload.php:41]: This circular dependency seems completely unnecessary.

File: src/command/parameter/Parameter.php [hunk 2]
```diff
@@ -0,0 +1,121 @@
...
+
+	/**
+	 * Returns the name of parameter
+	 */
+	abstract public function getTargetName() : string;
+
+	/**
+	 * Returns the translation of fail message
+	 */
+	public function getFailMessage(CommandSender $sender) : string{
```
Comment (dktapps) [inline src/command/parameter/Parameter.php:79]: Better to pass a `Language` instance in here directly. There's no need to depend on `CommandSender` (nor `Server`).

File: src/command/parameter/Parameter.php [hunk 3]
```diff
@@ -0,0 +1,121 @@
...
+	}
+
+	public function getPostfix() : ?string{
+		return $this->postfix;
+	}
+
+	public function prepare() : void{
+	}
+
+	public function setLength(int $length) : self{
```
Comment (dktapps) [inline src/command/parameter/Parameter.php:104]: what exactly is the purpose of this?

File: src/command/parameter/Parameter.php [hunk 4]
```diff
@@ -0,0 +1,121 @@
...
+	public function getEnum() : ?CommandEnum{
+		return $this->enum;
+	}
+
+	public function setPostfix(?string $postfix) : Parameter{
+		$this->postfix = $postfix;
+		return $this;
+	}
+
+	public function getPostfix() : ?string{
```
Comment (dktapps) [inline src/command/parameter/Parameter.php:97]: Postfixes only apply to integer parameters. Why is this here?

File: src/command/parameter/Parameter.php [hunk 5]
```diff
@@ -0,0 +1,121 @@
...
+	abstract public function getTargetName() : string;
+
+	/**
+	 * Returns the translation of fail message
+	 */
+	public function getFailMessage(CommandSender $sender) : string{
+		return $sender->getServer()->getLanguage()->translateString("%commands.generic.usage", [$this->getOverload()->getCommand()->getUsage()]);
+	}
+
+	public function setEnum(?CommandEnum $enum) : self{
```
Comment (dktapps) [inline src/command/parameter/Parameter.php:83]: Again, I don't see why this is embedded directly into the base `Parameter`. Also, the protocol types should not be exposed to the API.

File: src/command/parameter/defaults/IntegerParameter.php [hunk 6]
```diff
@@ -0,0 +1,42 @@
...
+*/
+
+declare(strict_types=1);
+
+namespace pocketmine\command\parameter\defaults;
+
+use pocketmine\command\CommandSender;
+use pocketmine\network\mcpe\protocol\AvailableCommandsPacket;
+
+class IntegerParameter extends FloatParameter{
```
Comment (dktapps) [inline src/command/parameter/defaults/IntegerParameter.php:29]: This is misleading. `instanceof FloatParameter` will return `true` in this case.

File: src/player/Player.php [hunk 7]
```diff
@@ -832,6 +832,10 @@ public function doFirstSpawn() : void{
 		if($this->getHealth() <= 0){
 			$this->respawn();
 		}
+
+		foreach($this->server->getOnlinePlayers() as $player){
+			$player->getNetworkSession()->syncAvailableCommands();
+		}
```
Comment (dktapps) [inline src/player/Player.php:838]: what is the purpose of this?

File: src/player/Player.php [hunk 8]
```diff
@@ -1973,6 +1977,10 @@ public function disconnect(string $reason, $quitMessage = null, bool $notify = t
 		}
 		$this->save();
 
+		foreach($this->server->getOnlinePlayers() as $player){
+			$player->getNetworkSession()->syncAvailableCommands();
+		}
```
Comment (dktapps) [inline src/player/Player.php:1982]: what is the purpose of this?

**Review by dktapps — CHANGES_REQUESTED** [review]
I haven't reviewed this in detail yet.

**Comment by lukeeey** [comment]
> It would be preferable to have callbacks bound to each overload instead of a central `onRun()`. That's how I'd personally planned to implement this.

That is preferable in most cases but that is a big change to how everyone is used to commands being handled. I think the way the PR implements it right now isn’t that bad. 

Maybe a way to optionally have callbacks if they wish would be better? I say this because when I was implementing brigadier into Nukkit (and brigadier also uses a sort of callback system but iirc it’s optional and you can just use a single method to handle everything), I ran into the fact that with the /title command you have options title, subtitle and actionbar. These things are all pretty much the same (they accept a text parameter) they just call different methods. 

With the callback style, I had three different methods with some duplicated code (sanity checks and whatnot) to do these simple tasks but I ended up making it shorter by putting them all into the same method and using a switch.

**Comment by dktapps** [comment]
@lukeeey it's possible to introduce some kind of legacy mapping which just registers `Command::execute()` as the only overload of a command to preserve backwards compatibility.

File: src/command/parameter/Parameter.php (hunk 4 above)
Comment (alvin0319) [reply src/command/parameter/Parameter.php:97]: https://github.com/pmmp/PocketMine-MP/blob/master/src/network/mcpe/protocol/types/command/CommandParameter.php#L43

If you look at the explanation here, postfix can only be set to null and string.

**Comment by lukeeey** [comment]
@dktapps That sounds like a good idea 👍

File: src/command/Overload.php [hunk 9]
```diff
@@ -0,0 +1,107 @@
...
+	public function getCommand() : Command{
+		return $this->command;
+	}
+
+	/**
+	 * @return Parameter[]
+	 */
+	public function getParameters() : array{
+		if(count($this->parameters) === 0){
+			return [new TextParameter("args", true)]; //Prevents parameters from being sent with empty values
```
Comment (lukeeey) [inline src/command/Overload.php:66]: But some commands may not have any parameters? As far as i am aware you can send a command to the client without any parameters and it will display just fine.

Comment (lukeeey) [reply src/command/Overload.php:66]: I may be understanding this wrong, actually. I'm guessing this wont be sent if there are no overloads added to the command.

File: src/player/Player.php (hunk 7 above)
Comment (alvin0319) [reply src/player/Player.php:838]: PlayerParameter needs to synchronize every time the players join or leaves.

Comment (lukeeey) [reply src/player/Player.php:838]: But since the type is `target` the client will show a list of players for you. This is the same with the `Item` enum name, which will show a list of items available on the client (although BDS does actually send the entire item list for some reason, you don't actually need to)

**Comment by alvin0319** [comment]
Fixed it by review, Please check the commits.

**Review by lukeeey — COMMENTED** [review]
An optional parameter doesn't need a default value... if the player has provided the parameter, it shouldn't return anything (because it doesnt exist)

File: src/command/Command.php [hunk 10]
```diff
@@ -101,6 +101,7 @@ public function execute(CommandSender $sender, string $commandLabel, array $args
 			return false;
 		}
 		foreach($this->overloads as $overload){
+			var_dump($overload->canParse($sender, $args));
```
Comment (SOF3) [inline src/command/Command.php]: remove this?

**Comment by alvin0319** [comment]
Is it okay to add `CommandSender` and `$params(array)` parameters to `Parameter->getFailMessage()` so that they can be converted from `Language`?

**Comment by dktapps** [comment]
Why not just take `Language` as a parameter?

**Comment by alvin0319** [comment]
Currently I used `Language` as a parameter, but something like `commands.generic.usage` do not require additional arguments?

**Comment by dktapps** [comment]
I don't understand what you're talking about.
